// rpcls/manager.js
var Referee = require('rpcls/referee').Referee;
var Client = require('rpcls/client').Client;
var utils = require('rpcls/utils');

var getClientId = utils.idGenerator();
var getGameId = utils.idGenerator();

function Manager(id) {
  this.sockets = {};
  this.games = {};
}

Manager.prototype = {
  registerClient: function(socket) {
    var client = new Client(this, socket, getClientId());
    this.sockets[client.id] = client.socket;
    ++this.nbOfSockets;
    
    console.info("new client(%d) connected", client.id);
  },

  unregisterClient: function(client) {
    --this.nbOfSockets;
    delete this.sockets[client.id];

    console.info('client(%d) disconnected. %d remaining', client.id, this.nbOfSockets);
  },

  createGame: function(player1) {
    var gameId = getGameId();
    var newGame = new Referee(gameId);
    this.games[newGame.gameId] = newGame;

    newGame.setPlayer1(player1);
  },

  joinGame: function(gameId, player2) {
    var existingGame = this.games[gameId];

    existingGame.setPlayer2(player2);
    existingGame.startGame();
  }
};

// Exports
exports.Manager = Manager;
