// rpcls/referee.js
var game = require('rpcls/game');

function Referee(gameId) {
  this.gameId = gameId;

  this.sign1 = null;
  this.sign2 = null;
}

Referee.prototype = {
  setPlayer1: function(player1) {
    this.player1 = player1;
    player1.referee = this;

    player1.gameCreated(this.gameId);
  },

  setPlayer2: function(player2) {
    this.player2 = player2;
    player2.referee = this;

    player2.gameJoined();
  },

  startGame: function() {
    this.game = new game.Game(this.player1, this.player2, 3);

    this.askToPlay();
  },

  askToPlay: function() {
    this.player1.askToPlay();
    this.player2.askToPlay();
  },

  play: function(player, sign) {
    if (player == this.player1) {
      this.sign1 = sign;
    } else if (player == this.player2) {
      this.sign2 = sign;
    } else {
      // Issue
    }
    this.ruleSet();
  },

  ruleSet: function() {
    if (null != this.sign1 && null != this.sign2) {
      var setWinner = this.game.play(this.sign1, this.sign2);

      if (this.game.hasWinner() && !this.endNotified) {
        this.endNotified = true;

        this.game.winner.win();
        this.game.loser.lose();
      }

      if (setWinner == this.player1) {
        this.player1.winSet();
        this.player2.loseSet();
      } else if (setWinner == this.player2) {
        this.player2.winSet();
        this.player1.loseSet();
      } else {
        this.player1.deuce();
        this.player2.deuce();
      }

      this.sign1 = null;
      this.sign2 = null;

      this.askToPlay();
    }
  },

  end: function() {
    this.player1.endGame();
    this.player2.endGame();
  }
};

// Exports
exports.Referee = Referee;
